File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.c
int lengthOfLongestSubstring(char* s) {
  int len = strlen(s);
  if(len == 0){
    return 0;
  }

  int last_seen[128];
  memset(last_seen , -1 , sizeof(last_seen));

  int max_len = 0; 
  int left = 0;

  for(int right = 0; right < len ; right++){
    char current_char = s[right];

    if(last_seen[current_char] >= left){
        left = last_seen[current_char] + 1;
    }

    last_seen[current_char] = right;

    int current_window_size = right - left + 1;
    if(current_window_size > max_len){
        max_len = current_window_size;
    }
  }

  return max_len;
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last_seen = {}
        max_len = 0
        left = 0

        for right , curr_char in enumerate(s):
            if curr_char in last_seen and last_seen[curr_char] >= left:
                left = last_seen[curr_char] + 1
            last_seen[curr_char] = right
            max_len = max(max_len , right - left + 1)
        return max_len